# Exploring Ebay Car Sales



## Introduction 

We'll work with a dataset of used cars from _eBay Kleinanzeigen_, a [classified](https://en.wikipedia.org/wiki/Classified_advertising) sections the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The data have been made a few modifications from the original dataset that was uploaded to Kaggle:
- We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
- We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)
***
The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.


The aim of this project is to clean the data and analyze the included used car listings.   
*** 




In [79]:
import pandas as pd 
import numpy as np 

autos = pd.read_csv("autos.csv", encoding=("Latin-1"))

In [8]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset contains 20 columns, most of which are strings. Some columns contain null-values, but there are no columns contain more than ~20% null value.

Now we'll start cleaning column names to make our next observations easier

***
## Cleaning Column Names
    
   The column name uses [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), which means we can't just replace space with underscore. 
   
   Let's convert the colum names from __camelcase__ to __snakecase__ and reword same column names based on dictionary to be more descriptive. 

In [64]:
## The column attribute to print an array of the existing column names.
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [81]:
## Assigning the modified column names back to the data frame column attribute
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [4]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

*** 
## Initial Exploring and Cleaning

Let's do some basic data exploration to see what other cleaning tasks need to be done. Initially we will look for : 
- Text columns where all or almost values are the same. These can often be dropped as they don't have useful information for analysis
- Numeric data stored as text value can be cleaned and converted

In [8]:
autos.describe(include='all')

date_crawled         name  price abtest vehicle_type  \
count                 50000        50000  50000  50000        44905   
unique                48213        38754   2357      2            8   
top     2016-03-21 20:37:19  Ford_Fiesta     $0   test    limousine   
freq                      3           78   1421  25756        12859   
mean                    NaN          NaN    NaN    NaN          NaN   
std                     NaN          NaN    NaN    NaN          NaN   
min                     NaN          NaN    NaN    NaN          NaN   
25%                     NaN          NaN    NaN    NaN          NaN   
50%                     NaN          NaN    NaN    NaN          NaN   
75%                     NaN          NaN    NaN    NaN          NaN   
max                     NaN          NaN    NaN    NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

As we can see above, `price` and `odometer` have numeric data stored as text that can be cleaned and converted. Need to do more investigations on `nr_of_pictures` column. `seller` and `offer_type` only have mostly one value as can be seen from __frequency row__. 

In [3]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

It seems the `nr_of_pictures` only have 0 values.

In [82]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)
autos.describe(include='all')

date_crawled         name  price abtest vehicle_type  \
count                 50000        50000  50000  50000        44905   
unique                48213        38754   2357      2            8   
top     2016-03-23 18:39:34  Ford_Fiesta     $0   test    limousine   
freq                      3           78   1421  25756        12859   
mean                    NaN          NaN    NaN    NaN          NaN   
std                     NaN          NaN    NaN    NaN          NaN   
min                     NaN          NaN    NaN    NaN          NaN   
25%                     NaN          NaN    NaN    NaN          NaN   
50%                     NaN          NaN    NaN    NaN          NaN   
75%                     NaN          NaN    NaN    NaN          NaN   
max                     NaN          NaN    NaN    NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

There are 2 columns, `price` and `odometer`, that stored numeric value as text. We will convert the data into numeric value so that we can do further analysis using the data.

In [83]:
autos["price"] = (autos["price"]
                  .str.replace("$", "")
                  .str.replace(",","")
                  .astype(int)
                 )

autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [84]:
autos["odometer"] = (autos["odometer"]
                    .str.replace("km", "")
                    .str.replace(",","")
                    .astype(int)
                    )
autos["odometer"].head()



0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [87]:
# Changing the odometer column name 
autos = autos.rename(columns={"odometer" : "odometer_km"}, inplace=False)

In [33]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer  registration_month fuel_type       brand  \
0  andere    150000                   3       lpg     peugeot   
1     7er    150000                   6    benzin         bmw   
2    golf     70000                   7    benzin  volkswagen   
3  fortwo     70000                   6    benzin       smart   
4   focus    150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Let's continue exploring the data, specifically looking for the data that doesn't feel right. We'll start by analyzing the `odometer_km` and `price` columns. Here's the steps we'll take:

- Analyze the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove



In [37]:
autos["price"].unique().shape[0]

2357

In [89]:
pd.set_option('float_format', '{:f}'.format) #To show the whole format number 

In [10]:
autos["price"].describe()

count      50000.000000
mean        9840.043760
std       481104.380500
min            0.000000
25%         1100.000000
50%         2950.000000
75%         7200.000000
max     99999999.000000
Name: price, dtype: float64

In [13]:
autos["price"].value_counts().head(10)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64

In [14]:
autos["price"].value_counts().sort_index(ascending = True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

In [16]:
autos["price"].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

Removing outliers -> that has 0 value and 99999999. 
The important data has value between 1 and 350000. 

From 350000 to 999990 and above is the way too significant. 

In [90]:
# Removing the outliers
autos = autos[autos["price"].between(1, 350000)]

***
## Exploring the date columns

Calculating the distribution of  `date_crawled`, `ad_created`, `last_seen`

In [15]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

#### Date Crawled

In [24]:
## This is date crawled series
date_crawled = autos["date_crawled"].str[:10]
date_crawled.describe()

count          48565
unique            34
top       2016-04-03
freq            1875
Name: date_crawled, dtype: object

In [25]:
(date_crawled
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05   0.025327
2016-03-06   0.014043
2016-03-07   0.036014
2016-03-08   0.033296
2016-03-09   0.033090
2016-03-10   0.032184
2016-03-11   0.032575
2016-03-12   0.036920
2016-03-13   0.015670
2016-03-14   0.036549
2016-03-15   0.034284
2016-03-16   0.029610
2016-03-17   0.031628
2016-03-18   0.012911
2016-03-19   0.034778
2016-03-20   0.037887
2016-03-21   0.037373
2016-03-22   0.032987
2016-03-23   0.032225
2016-03-24   0.029342
2016-03-25   0.031607
2016-03-26   0.032204
2016-03-27   0.031092
2016-03-28   0.034860
2016-03-29   0.034099
2016-03-30   0.033687
2016-03-31   0.031834
2016-04-01   0.033687
2016-04-02   0.035478
2016-04-03   0.038608
2016-04-04   0.036487
2016-04-05   0.013096
2016-04-06   0.003171
2016-04-07   0.001400
Name: date_crawled, dtype: float64

In [26]:
(date_crawled
 .value_counts(normalize=True, dropna=False)
 .sort_values(ascending=False)
)

2016-04-03   0.038608
2016-03-20   0.037887
2016-03-21   0.037373
2016-03-12   0.036920
2016-03-14   0.036549
2016-04-04   0.036487
2016-03-07   0.036014
2016-04-02   0.035478
2016-03-28   0.034860
2016-03-19   0.034778
2016-03-15   0.034284
2016-03-29   0.034099
2016-03-30   0.033687
2016-04-01   0.033687
2016-03-08   0.033296
2016-03-09   0.033090
2016-03-22   0.032987
2016-03-11   0.032575
2016-03-23   0.032225
2016-03-26   0.032204
2016-03-10   0.032184
2016-03-31   0.031834
2016-03-17   0.031628
2016-03-25   0.031607
2016-03-27   0.031092
2016-03-16   0.029610
2016-03-24   0.029342
2016-03-05   0.025327
2016-03-13   0.015670
2016-03-06   0.014043
2016-04-05   0.013096
2016-03-18   0.012911
2016-04-06   0.003171
2016-04-07   0.001400
Name: date_crawled, dtype: float64

#### ad created

In [13]:
ad_created = autos["ad_created"].str[:10]
ad_created.describe()

count          48565
unique            76
top       2016-04-03
freq            1887
Name: ad_created, dtype: object

In [28]:
(ad_created
.value_counts(normalize=True, dropna=False)
.sort_index()
)

2015-06-11   0.000021
2015-08-10   0.000021
2015-09-09   0.000021
2015-11-10   0.000021
2015-12-05   0.000021
2015-12-30   0.000021
2016-01-03   0.000021
2016-01-07   0.000021
2016-01-10   0.000041
2016-01-13   0.000021
2016-01-14   0.000021
2016-01-16   0.000021
2016-01-22   0.000021
2016-01-27   0.000062
2016-01-29   0.000021
2016-02-01   0.000021
2016-02-02   0.000041
2016-02-05   0.000041
2016-02-07   0.000021
2016-02-08   0.000021
2016-02-09   0.000021
2016-02-11   0.000021
2016-02-12   0.000041
2016-02-14   0.000041
2016-02-16   0.000021
2016-02-17   0.000021
2016-02-18   0.000041
2016-02-19   0.000062
2016-02-20   0.000041
2016-02-21   0.000062
               ...   
2016-03-09   0.033151
2016-03-10   0.031895
2016-03-11   0.032904
2016-03-12   0.036755
2016-03-13   0.017008
2016-03-14   0.035190
2016-03-15   0.034016
2016-03-16   0.030125
2016-03-17   0.031278
2016-03-18   0.013590
2016-03-19   0.033687
2016-03-20   0.037949
2016-03-21   0.037579
2016-03-22   0.032801
2016-03-23

In [29]:
(ad_created
.value_counts(normalize = True, dropna=False)
.sort_values(ascending=False)
)

2016-04-03   0.038855
2016-03-20   0.037949
2016-03-21   0.037579
2016-04-04   0.036858
2016-03-12   0.036755
2016-03-14   0.035190
2016-04-02   0.035149
2016-03-28   0.034984
2016-03-07   0.034737
2016-03-29   0.034037
2016-03-15   0.034016
2016-04-01   0.033687
2016-03-19   0.033687
2016-03-30   0.033501
2016-03-08   0.033316
2016-03-09   0.033151
2016-03-11   0.032904
2016-03-22   0.032801
2016-03-26   0.032266
2016-03-23   0.032060
2016-03-10   0.031895
2016-03-31   0.031875
2016-03-25   0.031751
2016-03-17   0.031278
2016-03-27   0.030989
2016-03-16   0.030125
2016-03-24   0.029280
2016-03-05   0.022897
2016-03-13   0.017008
2016-03-06   0.015320
               ...   
2016-02-20   0.000041
2016-02-12   0.000041
2016-02-26   0.000041
2016-02-18   0.000041
2016-02-14   0.000041
2016-01-10   0.000041
2016-02-24   0.000041
2016-02-05   0.000041
2016-02-02   0.000041
2016-02-16   0.000021
2016-01-16   0.000021
2016-02-17   0.000021
2016-02-22   0.000021
2015-12-05   0.000021
2016-01-22

In [18]:
last_seen = autos["last_seen"].str[:10]
last_seen.describe()

count          48565
unique            34
top       2016-04-06
freq           10772
Name: last_seen, dtype: object

In [34]:
(last_seen
 .value_counts(normalize=True, dropna=False)
 .sort_values())

2016-03-05   0.001071
2016-03-06   0.004324
2016-03-07   0.005395
2016-03-18   0.007351
2016-03-08   0.007413
2016-03-13   0.008895
2016-03-09   0.009595
2016-03-10   0.010666
2016-03-11   0.012375
2016-03-14   0.012602
2016-03-27   0.015649
2016-03-19   0.015834
2016-03-15   0.015876
2016-03-16   0.016452
2016-03-26   0.016802
2016-03-23   0.018532
2016-03-25   0.019211
2016-03-24   0.019767
2016-03-21   0.020632
2016-03-20   0.020653
2016-03-28   0.020859
2016-03-22   0.021373
2016-03-29   0.022341
2016-04-01   0.022794
2016-03-12   0.023783
2016-03-31   0.023783
2016-04-04   0.024483
2016-03-30   0.024771
2016-04-02   0.024915
2016-04-03   0.025203
2016-03-17   0.028086
2016-04-05   0.124761
2016-04-07   0.131947
2016-04-06   0.221806
Name: last_seen, dtype: float64

#### registration_year

In [20]:
registration_year = autos["registration_year"]

In [21]:
registration_year.describe()

count   48565.000000
mean     2004.755421
std        88.643887
min      1000.000000
25%      1999.000000
50%      2004.000000
75%      2008.000000
max      9999.000000
Name: registration_year, dtype: float64

The minimum is 1000 which is too far before the car was first invented

The maximum is 9999 which is many years into future

I decide the minimum year will be 1885 because that was the year the first car was invented. 

The maximum year will be 2016 because the year has many ads of cars

In [91]:
autos = autos[autos["registration_year"].between(1900, 2016)]


In [83]:
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_values(ascending=False).head(10)

2000   0.069834
2005   0.062776
1999   0.062464
2004   0.056988
2003   0.056779
2006   0.056384
2001   0.056280
2002   0.052740
1998   0.051074
2007   0.047972
Name: registration_year, dtype: float64

It appears that the most cars were registered in the past 20 years

## Exploring Price by Brand

In [27]:
## This is the top 20

(autos["brand"]
 .value_counts(normalize=True)
 .sort_values(ascending=False)
 .head(20))

volkswagen       0.213740
opel             0.109220
bmw              0.108580
mercedes_benz    0.094680
audi             0.085660
ford             0.069580
renault          0.048080
peugeot          0.029120
fiat             0.026160
seat             0.018820
skoda            0.015720
mazda            0.015140
nissan           0.015080
smart            0.014020
citroen          0.014020
toyota           0.012340
sonstige_autos   0.010920
hyundai          0.009760
volvo            0.009140
mini             0.008480
Name: brand, dtype: float64

In [8]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brand = (autos["brand"].value_counts(normalize=True))[brand_counts > 0.05].index

I don't really see any special pattern on the top 20 list cars, besides I prefer to select those that have a certatin percentage of the total (e.g. > 5%)

In [9]:
for brand in common_brand:
    print(brand)

volkswagen
opel
bmw
mercedes_benz
audi
ford


In [92]:
brand_mean_prices = {}

for brand in common_brand:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

- Audi, BMW, Mercedes Bens are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between 

In [93]:
bmp_series = pd.Series(brand_mean_prices)
bmp_series

audi             9336
bmw              8332
ford             3749
mercedes_benz    8628
opel             2975
volkswagen       5402
dtype: int64

In [94]:
DF = pd.DataFrame(bmp_series, columns=["mean_price"])
DF

mean_price
audi                 9336
bmw                  8332
ford                 3749
mercedes_benz        8628
opel                 2975
volkswagen           5402

In [97]:
brand_mean_milleage = {}

for brand in common_brand: 
    brand_only = autos[autos["brand"] == brand]
    milleage_mean = brand_only["odometer_km"].mean()
    brand_mean_milleage[brand] = int(milleage_mean)

brand_mean_milleage 

{'audi': 129157,
 'bmw': 132572,
 'ford': 124266,
 'mercedes_benz': 130788,
 'opel': 129310,
 'volkswagen': 128707}

In [99]:
bmm = pd.Series(brand_mean_milleage)
DF["mean_milleage"] = bmm
DF

mean_price  mean_milleage
audi                 9336         129157
bmw                  8332         132572
ford                 3749         124266
mercedes_benz        8628         130788
opel                 2975         129310
volkswagen           5402         128707